### PR TITLE
[FLINK-35543][HIVE] Upgrade Hive 2.3 connector to version 2.3.10

### PR DESCRIPTION
**Pull request by pan3793**
## What is the purpose of the change

Hive 2.3.10 is the final release of branch-2.3, which focuses on mitigating CVE dependencies.

https://github.com/apache/hive/blob/rel/release-2.3.10/RELEASE_NOTES.txt

Spark upgraded Hive 2.3.10 in 4.0.0-preview1 via SPARK-47018

## Brief change log

- Upgrade `hive.version` from 2.3.9 to 2.3.10
- Introduce a `HiveShimV2310`
- Rename module `flink-sql-connector-hive-2.3.9` to `flink-sql-connector-hive-2.3.10`
- HIVE-25098 upgrades Thrift from 0.9 to 0.14, and THRIFT-5297 (0.14.0) removes `requestTimeout` and `beBackoffSlotLength`, code are updated correspondingly on the Flink SQL Gateway
- Update LICENSE/NOTICE accordingly, please refer to release notes for details
- Update docs to match the code

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs are updated accordingly

### CODE REVIEW

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/endpoint/hive/HiveServer2Endpoint.java
Comment (JingGe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/endpoint/hive/HiveServer2Endpoint.java:872]: could you add clarification in the PR description for the reason of removing those code?

Comment (pan3793) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/endpoint/hive/HiveServer2Endpoint.java:872]: methods and args were removed in new thrift versions. you can check HIVE-25098 for details.

Comment (pan3793) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/endpoint/hive/HiveServer2Endpoint.java:872]: please leave review comments inline, I will summarize all conversions into the PR description eventually.

Comment (JingGe) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/endpoint/hive/HiveServer2Endpoint.java:872]: Thanks for the clarification.

**Review by JingGe — APPROVED** [review]
LGTM

**Review by RocMarshal — APPROVED** [review]
LGTM +1.

**Comment by pan3793** [comment]
OKay, CI is green, and PR description is updated, should be ready to go.

**Comment by pan3793** [comment]
ping @JingGe, is there any thing I can do before merging?

**Comment by pan3793** [comment]
ping @1996fanrui can we include this in 1.20?

**Review by 1996fanrui — APPROVED** [review]
> ping @1996fanrui can we include this in 1.20?

Thanks @pan3793 for the ping. I saw the core change is upgrading the hive version from `2.3.9` to `2.3.10`, it only upgrades one patch version, and I checked the release note for 2.3.10. I think it doesn't introduce any risk, so LGTM.

I will merge it to 1.20 (due to this PR already has 2 approvals) if no objection tomorrow.

**Comment by pan3793** [comment]
thank you, @1996fanrui

**Pull request merged**
